File: solve-x-main/backend/models/application.go
package models

import "gorm.io/gorm"

type Application struct {
	gorm.Model
	StudentID     uint         `json:"student_id" gorm:"not null"`
	Student       *Student     `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;foreignKey:StudentID;references:ID"`
	OpportunityID uint         `json:"opportunity_id" gorm:"not null"`
	Opportunity   *Opportunity `gorm:"constraint:OnUpdate:CASCADE,OnDelete:CASCADE;foreignKey:OpportunityID;references:ID"`
	Message       string       `json:"message" gorm:"type:text"`
	ResumeLink    string       `json:"resume_link"`
	Status        string       `json:"status" gorm:"type:TEXT CHECK(status IN ('pending','accepted','rejected'));not null;default:'pending'"`
}

const (
	StatusPending  = "pending"
	StatusAccepted = "accepted"
	StatusRejected = "rejected"
)

func (s Student) CreateApplication(db *gorm.DB, opportunityID uint, message, resumeLink string) error {
	application := Application{
		StudentID:     s.ID,
		OpportunityID: opportunityID,
		Message:       message,
		ResumeLink:    resumeLink,
		Status:        StatusPending,
	}
	return db.Create(&application).Error
}

func (s Student) DeleteApplication(db *gorm.DB, opportunityID uint) error {
	return db.
		Where("student_id = ? AND opportunity_id = ?", s.ID, opportunityID).
		Delete(&Application{}).
		Error
}

func GetApplicationsByOpportunityID(db *gorm.DB, opportunityID uint) ([]Application, error) {
	var applications []Application
	err := db.
		Preload("Student").
		Preload("Opportunity").
		Where("opportunity_id = ?", opportunityID).
		Find(&applications).
		Error
	return applications, err
}

// GetApplicationsByStudentID returns all applications submitted by a student
func GetApplicationsByStudentID(db *gorm.DB, studentID uint) ([]Application, error) {
	var applications []Application
	err := db.
		Preload("Student").
		Preload("Opportunity").
		Preload("Opportunity.Professor").
		Where("student_id = ?", studentID).
		Order("created_at DESC").
		Find(&applications).
		Error
	return applications, err
}

// GetApplicationsByProfessorOpportunities returns all applications for opportunities created by a professor
func GetApplicationsByProfessorOpportunities(db *gorm.DB, professorID uint) ([]Application, error) {
	var applications []Application
	err := db.
		Preload("Student").
		Preload("Opportunity").
		Preload("Opportunity.Professor").
		Joins("JOIN opportunities ON opportunities.id = applications.opportunity_id").
		Where("opportunities.professor_id = ?", professorID).
		Order("applications.created_at DESC").
		Find(&applications).
		Error
	return applications, err
}

// GetApplicationByID returns an application by ID
func GetApplicationByID(db *gorm.DB, id uint) (*Application, error) {
	var application Application
	err := db.
		Preload("Student").
		Preload("Opportunity").
		Preload("Opportunity.Professor").
		First(&application, id).
		Error
	return &application, err
}

// UpdateApplicationStatus updates the status of an application and notifies the student
func UpdateApplicationStatus(db *gorm.DB, id uint, status string) (*Application, error) {
	// Validate status
	if status != StatusPending && status != StatusAccepted && status != StatusRejected {
		return nil, gorm.ErrInvalidValue
	}

	application, err := GetApplicationByID(db, id)
	if err != nil {
		return nil, err
	}

	oldStatus := application.Status
	application.Status = status
	if err := db.Save(application).Error; err != nil {
		return nil, err
	}

	// Notify student if status changed to accepted or rejected
	if oldStatus != status && (status == StatusAccepted || status == StatusRejected) {
		opportunityName := "this opportunity"
		if application.Opportunity != nil {
			opportunityName = application.Opportunity.Name
		}
		NotifyApplicationStatusChange(db, application.StudentID, opportunityName, status)
	}

	return application, nil
}

File: lib/services/application.ts
import prisma from "@/lib/prisma";
import { createNotification } from "./notification";

export type ApplicationStatus = "pending" | "accepted" | "rejected";

export const STATUS_PENDING = "pending" as const;
export const STATUS_ACCEPTED = "accepted" as const;
export const STATUS_REJECTED = "rejected" as const;

/**
 * Create an application for a student
 */
export async function createApplication(
  applicantType: "STUDENT" | "PROFESSOR",
  applicantId: bigint,
  opportunityId: bigint,
  message?: string,
  resumeLink?: string
): Promise<any> {
  // Check opportunity and constraints
  const opportunity = await prisma.opportunity.findUnique({ where: { id: opportunityId } });
  if (!opportunity) throw new Error("Opportunity not found");

  // Professors may only apply to research opportunities
  if (applicantType === "PROFESSOR" && opportunity.type !== "research") {
    throw new Error("Professors can only apply to research opportunities");
  }

  // Check if application already exists for this applicant
  const existing = await prisma.application.findFirst({
    where: (applicantType === "STUDENT"
      ? { studentId: applicantId, opportunityId }
      : { professorApplicantId: applicantId, opportunityId }) as any,
  } as any);

  if (existing) {
    throw new Error("You have already applied to this opportunity");
  }

  const data: any = {
    opportunityId,
    message: message || null,
    resumeLink: resumeLink || null,
    status: STATUS_PENDING,
    applicantType: applicantType,
  };
  if (applicantType === "STUDENT") data.studentId = applicantId;
  else data.professorApplicantId = applicantId;

  try {
    const application = await prisma.application.create({
      data,
      include: {
        student: true,
        professorApplicant: true,
        opportunity: {
          include: {
            professor: true,
            student: true,
          },
        },
      },
    } as any);

    return application;
  } catch (err: any) {
    console.error("prisma.application.create failed", { err: err?.message || err });
    throw err;
  }
}

/**
 * Delete an application
 */
export async function deleteApplication(
  applicantType: "STUDENT" | "PROFESSOR",
  applicantId: bigint,
  opportunityId: bigint
): Promise<void> {
  const where: any = { opportunityId };
  if (applicantType === "STUDENT") where.studentId = applicantId;
  else where.professorApplicantId = applicantId;
  await prisma.application.deleteMany({ where: where as any } as any);
}

/**
 * Get applications by opportunity ID
 */
export async function getApplicationsByOpportunityId(opportunityId: bigint): Promise<any[]> {
  return prisma.application.findMany({
    where: { opportunityId } as any,
    include: {
      student: true,
      professorApplicant: true,
      opportunity: {
        include: { professor: true, student: true },
      },
    },
    orderBy: { createdAt: "desc" },
  } as any);
}

/**
 * Get applications by student ID
 */
export async function getApplicationsByStudentId(studentId: bigint): Promise<any[]> {
  return prisma.application.findMany({
    where: { studentId } as any,
    include: {
      student: true,
      professorApplicant: true,
      opportunity: {
        include: { professor: true, student: true },
      },
    },
    orderBy: { createdAt: "desc" },
  } as any);
}

/**
 * Get applications for all opportunities by a professor
 */
export async function getApplicationsByProfessorOpportunities(
  professorId: bigint
): Promise<any[]> {
  return prisma.application.findMany({
    where: {
      opportunity: {
        professorId,
      },
    } as any,
    include: {
      student: true,
      professorApplicant: true,
      opportunity: {
        include: { professor: true, student: true },
      },
    },
    orderBy: { createdAt: "desc" },
  } as any);
}

/**
 * Get applications for opportunities owned by a specific user.
 * For professors this returns applications for opportunities where opportunity.professorId = ownerId
 * For students this returns applications for opportunities where opportunity.studentId = ownerId
 */
export async function getApplicationsByOwnedOpportunities(
  ownerId: bigint,
  ownerRole: "professor" | "student"
): Promise<any[]> {
  if (ownerRole === "professor") {
    return getApplicationsByProfessorOpportunities(ownerId as bigint);
  }

  // student-owned opportunities
  return prisma.application.findMany({
    where: {
      opportunity: {
        studentId: ownerId,
      },
    } as any,
    include: {
      student: true,
      professorApplicant: true,
      opportunity: {
        include: { professor: true, student: true },
      },
    },
    orderBy: { createdAt: "desc" },
  } as any);
}

/**
 * Get application by ID
 */
export async function getApplicationById(id: bigint): Promise<any> {
  const application = await prisma.application.findUnique({
    where: { id } as any,
    include: {
      student: true,
      professorApplicant: true,
      opportunity: {
        include: { professor: true, student: true },
      },
    },
  } as any);

  if (!application) {
    throw new Error("Application not found");
  }

  return application as any;
}

/**
 * Get applications submitted by an applicant (student or professor)
 */
export async function getApplicationsByApplicant(applicantType: "STUDENT" | "PROFESSOR", applicantId: bigint): Promise<any[]> {
  const where: any = applicantType === "STUDENT" ? { studentId: applicantId } : { professorApplicantId: applicantId };
  return prisma.application.findMany({
    where: where as any,
    include: {
      student: true,
      professorApplicant: true,
      opportunity: { include: { professor: true, student: true } },
    },
    orderBy: { createdAt: "desc" },
  } as any);
}

/**
 * Update application status
 */
export async function updateApplicationStatus(
  id: bigint,
  status: ApplicationStatus
): Promise<any> {
  // Validate status
  if (![STATUS_PENDING, STATUS_ACCEPTED, STATUS_REJECTED].includes(status)) {
    throw new Error("Invalid status value");
  }

  const application = await getApplicationById(id);
  const oldStatus = application.status;

  const updated = await prisma.application.update({
    where: { id } as any,
    data: { status },
    include: {
      student: true,
      professorApplicant: true,
      opportunity: {
        include: { professor: true, student: true },
      },
    },
  } as any);

  // Notify student if status changed to accepted or rejected
  if (oldStatus !== status && (status === STATUS_ACCEPTED || status === STATUS_REJECTED)) {
    const updatedAny: any = updated;
    const opportunityName = updatedAny.opportunity?.name || "this opportunity";
    // Notify applicant (student or professor)
    if (updatedAny.applicantType === "STUDENT" && updatedAny.studentId) {
      await notifyApplicationStatusChange(updatedAny.studentId, opportunityName, status, "student");
    } else if (updatedAny.applicantType === "PROFESSOR" && updatedAny.professorApplicantId) {
      await notifyApplicationStatusChange(updatedAny.professorApplicantId, opportunityName, status, "professor");
    }
  }

  return updated as any;
}

/**
 * Helper to notify student of application status change
 */
async function notifyApplicationStatusChange(
  recipientId: bigint,
  opportunityName: string,
  status: string,
  recipientRole: "student" | "professor"
) {
  let title = "Application Update";
  let message = `Your application for '${opportunityName}' has been ${status}`;
  let type: "info" | "success" | "warning" | "error" = "info";

  if (status === STATUS_ACCEPTED) {
    title = "🎉 Application Accepted!";
    type = "success";
  } else if (status === STATUS_REJECTED) {
    message = `Your application for '${opportunityName}' was not accepted this time`;
  }

  await createNotification(recipientId, recipientRole, title, message, type);
}
